README.md
[![NPM version][npm-version-image]][npm-url] [![NPM downloads][npm-downloads-image]][npm-url] [![MIT License][license-image]][license-url]

A simple module for displaying days remaining before the scheduled event on a ribbon.

## [Demo](http://paveldemyanenko.github.io/ng-countdown-ribbon/index.html)

## [Changelog](CHANGELOG.md)

###Usage

After including *ng-countdown-ribbon.min.js* and *ng-countdown-ribbon.min.css*, inject the ng-countdown-ribbon provider into your project.

```javascript
var app = angular.module('demo', ['ngCountdownRibbon']);
```
Now you can trigger countdown ribbon from anywhere in your app.  To display a ribbon, just use the `set` method.

First parameter is the end date and second is for the event link. Use date format in [ISO 8601](https://xkcd.com/1179/).

```javascript
ngCountdownRibbon.set('2015-03-26', 'http://amzn.com/w/ZWFNUL8AGNLP');
```
## License

ng-countdown-ribbon is freely distributable under the terms of the [MIT license](LICENSE).

[license-image]: http://img.shields.io/badge/license-MIT-blue.svg?style=flat
[license-url]: LICENSE

[npm-url]: https://npmjs.org/package/ng-countdown-ribbon
[npm-version-image]: http://img.shields.io/npm/v/ng-countdown-ribbon.svg?style=flat
[npm-downloads-image]: http://img.shields.io/npm/dm/ng-countdown-ribbon.svg?style=flat



CHANGELOG.md
Changelog
=========

### 0.1.0

Initial release.

### 0.2.0

Add ability to add custom themes.

### 0.3.0

Add event types.

Gruntfile.js
module.exports = function(grunt) {
    'use strict';

    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-sass');
    grunt.loadNpmTasks('grunt-contrib-cssmin');
    grunt.loadNpmTasks('grunt-processhtml');
    grunt.loadNpmTasks('grunt-contrib-htmlmin');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-watch');

    grunt.initConfig({

        jshint: {
            files: [
                'Gruntfile.js',
                'src/scripts/**/*.js',
                'demo/**/*.js'
            ],
            options: {
                jshintrc: '.jshintrc'
            }
        },

        clean: {
            dist: ['dist']
        },

        sass: {
            build: {
                files: {
                    'src/styles/ng-countdown-ribbon.css': 'src/styles/ng-countdown-ribbon.scss'
                }
            }
        },

        cssmin: {
            build: {
                files: {
                    'dist/ng-countdown-ribbon.min.css': 'src/styles/ng-countdown-ribbon.css'
                }
            },

            demo: {
                files: {
                    '_gh-pages/demo.min.css': [
                        'src/styles/ng-countdown-ribbon.css',
                        'demo/demo.css'
                    ]
                }
            }
        },

        processhtml: {
            build: {
                files: {
                    '_gh-pages/index.html': 'demo/index.html'
                }
            }
        },

        htmlmin: {
            options: {
                removeComments: true,
                collapseWhitespace: true
            },
            index: {
                files: {
                    '_gh-pages/index.html': '_gh-pages/index.html'
                }
            }
        },

        uglify: {
            build: {
                files: {
                    'dist/ng-countdown-ribbon.min.js': 'src/**/*.js'
                }
            },
            demo: {
                files: {
                    '_gh-pages/demo.min.js': [
                        'bower_components/angular/angular.js',
                        'src/scripts/ng-countdown-ribbon.js',
                        'demo/demo.js'
                    ]
                }
            }
        },

        watch: {
            scripts: {
                files: ['src/**/*', 'demo/**/*'],
                tasks: ['build']
            }
        }

    });

    grunt.registerTask('build', [
        'jshint',
        'clean',
        'sass',
        'cssmin:build',
        'uglify:build',
        'demo'
    ]);

    grunt.registerTask('demo', [
        'processhtml',
        'htmlmin',
        'cssmin:demo',
        'uglify:demo'
    ]);

    grunt.registerTask('default', ['build']);
    grunt.registerTask('dev', ['build', 'watch']);
};


src/scripts/ng-countdown-ribbon.js
/**
 * @license ng-countdown-ribbon v0.1.0
 * http://paveldemyanenko.github.io/ng-countdown-ribbon
 * (c) 2015 MIT License, paveldemyanenko.com
 */
(function() {
    'use strict';

    /**
     * @description
     *
     * This module provides AngularJS application with a system for displaying
     * time remaining before the scheduled event on a ribbon.
     */

    var module = angular.module('ngCountdownRibbon', []);

    module.provider('ngCountdownRibbon', function() {

        this.$get = ['$document', '$compile', '$rootScope',

            function($document, $compile, $rootScope) {

                var options = {
                    theme: 'black',
                    position: 'left'
                };

                var themes = {
                    black: 'ribbon_black',
                    blue: 'ribbon_blue'
                };

                var positions = {
                    left: 'ribbon_left',
                    right: 'ribbon_right'
                };

                var types = {
                    birthday: 'birthday',
                    event: 'special-event'
                };

                var ribbonScope = $rootScope.$new();

                var tpl = $compile(
                  '<a ng-href="{{ngCountdownRibbon.ribbonLink}}" class="ribbon-container" target="_blank">' +
                  '<span class="ribbon" ng-class="ngCountdownRibbon.ribbonClass">' +
                  '<span><ng-pluralize count="ngCountdownRibbon.ribbonDays" ' +
                  'when="{\'0\': \'Event has come\', \'one\': \'1 day left\', \'other\': \'{} days left\'}"' +
                  '</ng-pluralize></span></span></a>'
                )(ribbonScope);

                $document.find('body').append(tpl);

                var setTheme = function(providedTheme) {
                    var theme = providedTheme || options.theme;
                    return themes[theme] || themes.black;
                };

                var setPosition = function(providedPosition) {
                    var position = providedPosition || options.position;
                    return positions[position] || positions.left;
                };

                var setType = function(providedType) {
                    var type = (providedType || options.type);
                    return types[type] || types.birthday;
                };


                var ribbonObject = {

                    config: function(params) {
                        params = params || {};
                        angular.extend(options, params);
                    },

                    set: function(date, link, userOpt) {

                        if (!date) {
                            return;
                        }

                        var endDate = new Date(date.split('-').join('/'));

                        var asUTC = function(date) {
                            var result = new Date(date);
                            result.setMinutes(result.getMinutes() - result.getTimezoneOffset());
                            return result;
                        };

                        var daysBetween = function(startDate, endDate) {
                            var millisecondsPerDay = 24 * 60 * 60 * 1000;
                            return (asUTC(endDate) - asUTC(startDate)) / millisecondsPerDay;
                        };

                        var currentDate = new Date().getTime();

                        var days = Math.round(daysBetween(currentDate, endDate));

                        var userOpts = {};

                        if (typeof userOpt === 'object') {
                            userOpts = {
                                theme: userOpt.theme || undefined,
                                position: userOpt.position || undefined,
                                type: userOpt.type || undefined
                            };
                        } else {
                            userOpts.type = userOpt;
                        }

                        var ribbonClass = setTheme(userOpts.theme) + ' ' + setPosition(userOpts.position) + ' ' + setType(userOpts.type);

                        ribbonScope.ngCountdownRibbon = {
                            ribbonLink: link,
                            ribbonClass: ribbonClass,
                            ribbonDays: days
                        };
                    },

                    addTheme: function(themeName, themeClass) {
                        if (!themeName || !themeClass) {
                            return;
                        }
                        themes[themeName] = themeClass;
                    },

                    addType: function(typeName, typeClass) {
                        if (!typeName || !typeClass) {
                            return;
                        }
                        types[typeName] = typeClass;
                    }
                };
                return ribbonObject;
            }
        ];
    });
})();


demo/demo.js
var app = angular.module('demo', ['ngCountdownRibbon']);

app.controller('MainCtrl', ['$scope', 'ngCountdownRibbon',
    function($scope, ngCountdownRibbon) {
        'use strict';

        $scope.theme = 'black';
        $scope.themeOptions = ['black', 'blue'];

        $scope.position = 'left';
        $scope.positionOptions = ['left', 'right'];

        ngCountdownRibbon.addTheme('greenTheme', 'green');

        ngCountdownRibbon.config({
            theme: 'greenTheme'
        });

        ngCountdownRibbon.set('2015-03-26', 'http://amzn.com/w/ZWFNUL8AGNLP');

        $scope.setDefaultPosition = function() {
            ngCountdownRibbon.config({
                position: $scope.position
            });
            ngCountdownRibbon.set('2015-03-26', 'http://amzn.com/w/ZWFNUL8AGNLP');
        };

        $scope.setDefaultTheme = function() {
            ngCountdownRibbon.config({
                theme: $scope.theme
            });
            ngCountdownRibbon.set('2015-03-26', 'http://amzn.com/w/ZWFNUL8AGNLP');
        };

    }
]);
